lib/p2p/Pool.ts

```diff
@@ -0,0 +1,172 @@
+import net, { Server, Socket } from 'net';
+import { EventEmitter } from 'events';
+import Peer from './Peer';
+import Hosts from './Hosts';
+import NetAddress from './NetAddress';
+import Logger from '../Logger';
+
+interface PoolConfig {
+  listen: boolean;
+  port: number;
+}
+
+/** A pool of peers for handling all network activity */
+class Pool extends EventEmitter {
+  private peers: { [ key: string ]: Peer };
+  private server: Server;
+  private logger: Logger;
+  private connected: boolean;
+  private hosts: Hosts;
+
+  constructor(private config: PoolConfig) {
+    super();
+    this.peers = {};
+    this.logger = Logger.global;
+    this.server = net.createServer();
+    this.hosts = new Hosts();
+    this.connected = false;
+
+    this._init();
+  }
+
+  public connect = async (): Promise<void> => {
+    if (this.connected) {
+      return;
+    }
+
+    await this.hosts.init();
+
+    // p2p bootstrap process here
+
+    this.listen();
+    this.connected = true;
+  }
+
+  public disconnect = async (): Promise<void> => {
+    if (!this.connected) {
+      return;
+    }
+
+    await this.unlisten();
+    this.destroyPeers();
+
+    this.connected = false;
+  }
+
+  public addOutbound = async (host: string, port: number): Promise<Peer> => {
+    const address = new NetAddress(host, port);
+    const peer = Peer.fromOutbound(address);
+
+    this.bindPeer(peer);
+
+    await peer.open();
+    this.addPeer(peer);
+
+    return peer;
+  }
+
+  public broadcastOrder = (order: any) => {
+    Object.keys(this.peers).map((key) => {
+      this.peers[key].sendOrder(order);
+    });
+  }
+
+  private _init = () => {
+    this.server.on('error', (err) => {
+      console.log('err: ' + err);
+    });
+
+    this.server.on('connection', (socket: Socket) => {
+      this.handleSocket(socket);
+    });
+
+    this.server.on('listening', () => {
+      const { address, port } = this.server.address();
+      this.logger.info(`pool server listening on ${address}:${port}`);
+    });
+  }
+
+  private addInbound = async (socket: Socket): Promise<Peer> => {
+    const peer = Peer.fromInbound(socket);
+
+    this.bindPeer(peer);
+
+    await peer.open();
+    this.addPeer(peer);
+
+    return peer;
+  }
+
+  private handleSocket = (socket: Socket) => {
+    if (!socket.remoteAddress) {
+      this.logger.debug('Ignoring disconnected peer');
+      socket.destroy();
+      return;
+    }
+
+    const host = socket.remoteAddress;
+    if (this.hosts.isBanned(host)) {
+      this.logger.debug(`Ignoring banned peer (${host})`);
+      socket.destroy();
+      return;
+    }
+
+    this.addInbound(socket);
+  }
+
+  private handlePacket = (packet) => {
+    // TODO: handle
+    console.log('packet: ' + JSON.stringify(packet));
+  }
+
+  private bindPeer = (peer: Peer) => {
+    peer.on('packet', (packet) => {
+      this.handlePacket(packet);
+    });
+
+    peer.on('error', (err) => {
+      this.logger.error('peer error', err);
```

Comment:
Yes, you're right. The error handling wasn't complete in this PR. i'll try to address at least that before we'll merge it.

Reply:
I removed the error event in case of parser error. It will be managed inside of `Peer`: log it, and increase ban score. `Pool` doesn't need to know about it, unless it reaches a `ban` event.

Reply:
That makes sense, but in that case wouldn't we want to call `removePeer` here since the socket error this handles will always result in a closed connection?